### ISSUE TITLE
Soltys (1995)

### ISSUE THREAD

**Issue by parricc** [exoscoriae/eXoDOS#5213]
The SoundBlaster option does not work correctly.
When choosing Port: 220, IRQ: 7, DMA: 1, there is no sound.
When choosing Auto, the game freezes after you press a key.

This bug happens in both Windows and Linux.
Tested with ECE r4230 in Windows.
Testing with ECE r4301, r4358, and r4482 in Linux.

**Comment by interloper98** [comment]
Tried w/ DOSBox Staging 0.81.1 with:

```ini
[cpu]
cycles = 15000

[sblaster]
sbtype = sb16

[gus]
gus = false
```

At least music & sfx work, and doesn't lock up:

https://github.com/exoscoriae/eXoDOS/assets/165201780/2a3c116e-7466-4d6b-8e34-ca447e784e9a

**Comment by SmilingSpectre** [comment]
Partially confirmed.

1. If SB=SB1 or SB2 music works as expected, but not FX. But for SBPRO1,2 or SB16 there is no sound at all for 240,7,1 (I forced it in config - no luck).
2. For any SB, if you select "Sound Blaster" in menu, then select "AUTO" for each of three choices, sound works. And music too.
3. If you _don't_ select "Sound Blaster", but select "Automatic Configuration", you get sound - but game stuck on the first screen, playing melody forever.

Tried to change the settings. On 210,7,1 all SB woks as SB1: music is here, but sound isn't. Again: everything works on "AUTO".

I recommend to send the issue to Staging guys. :)

**Comment by Python-Exoproject** [comment]
Whats happening is that with all the sound devices turned on as we do in current eXoDOS conf the game doesnt know what to do when you select "Automatic Configuration" and crashes. Turn all but the sound device being used off in the conf and auto detect works.

As for the FX not working unless Auto is selected, this is related to the DMA selection as you can specify the IRQ and Port using Auto only for DMA and it works. I lean towards a game issue as all forks perform the same, its detecting DMA properly by itself with auto but not working when we specify the DMA.

I think the best solution is to edit run.bat to tell the user to select "Auto" for each option, then everything is fine. I also noticed that when GUS was selected we had the fluidsynth turn on for some reason so ive fixed that as well. Here is a run.bat with the suggested changes: [run.zip](https://github.com/user-attachments/files/15524607/run.zip)

I also found on my PC when sound canvas is selected while using ECE there is noise in the channel but on Staging its clean. I would like to change to Staging for the main launch due to this however one problem that does seem Staging specific seems to be that it wont give sound for GUS choice. Can you confirm @SmilingSpectre?

**Comment by interloper98** [comment]
> one problem that does seem Staging specific seems to be that it wont give sound for GUS choice. 

Ran some tests with older releases, and between 0.75.2 and 0.76 it stopped working. 

It seemed like Solty's wasn't able to start the card playing in the latter versions.

Looking at the logs, normally Staging prints the number of GUS voices, but in this case, it was being reported.

I compared those areas of code and and with some trial and error on my RetroPie box, this was enough to get it playing again:

```diff
diff --git a/src/hardware/gus.cpp b/src/hardware/gus.cpp
index a46ff6ff1..7af628469 100644
--- a/src/hardware/gus.cpp
+++ b/src/hardware/gus.cpp
@@ -684,6 +684,7 @@ void Gus::ActivateVoices(uint8_t requested_voices)
                ms_per_render = MillisInSecond / sample_rate_hz;

                audio_channel->SetSampleRate(sample_rate_hz);
+               BeginPlayback();
        }
 }

@@ -769,8 +770,6 @@ void Gus::AudioCallback(const uint16_t num_requested_frames)

 void Gus::BeginPlayback()
 {
-       dac_enabled = ((register_data & 0x200) != 0);
-       irq_enabled = ((register_data & 0x400) != 0);

        audio_channel->Enable(true);

@@ -1570,8 +1569,9 @@ void Gus::WriteToRegister()
                        StopPlayback();
                } else if (state == 1) {
                        PrepareForPlayback();
-               } else if (active_voices) {
-                       BeginPlayback();
+               } else {
+                       dac_enabled = ((register_data & 0x200) != 0);
+                       irq_enabled = ((register_data & 0x400) != 0);
                }
        }
                return;
```

This is just a guess and will do more tests on my main win 11 environment later.

If tests are OK I'll pass it onto the Staging devs to see if they can do something with it.

**Comment by johnnovak** [comment]
Thanks @interloper98. That active voice check was questionable to begin with; hardware usually is "dumb" and works all the time. Sound cards certainly don't try to "micro optimise" their "work"; those logic circuits do 100% work 100% of the time 😏

**Comment by exoscoriae** [comment]
@Python-Exoproject I added your run.bat file to the zip.

**Comment by exoscoriae** [comment]
moving to low priority as we are now waiting on staging

**Comment by Python-Exoproject** [comment]
This is fixed in Staging and will be in the 82 release

**Comment by exoscoriae** [comment]
processed
